### PR TITLE
Provide slightly better C# experience

### PR DESCRIPTION
**Pull request by AlexeyRaga**
- More consistent naming for C# side (F# is not affected)
- An extension method `AutoGenConfig.WithGenerator` that is only visible from C# to provide better experience for creating autogen configs in C#

### CODE REVIEW

**Review by TysonMN — CHANGES_REQUESTED** [review]

File: src/Hedgehog.Experimental/Gen.fs [hunk 1]
```diff
@@ -116,6 +124,7 @@ module GenX =
     |> Gen.map (fun s -> "#" + s)
 
   /// Generates a random URI.
+  [<CompiledName("Uri")>]
```
[the remark this answers is not in the document]
Comment (TysonMN) [reply src/Hedgehog.Experimental/Gen.fs:127]: I think putting the F# API in one namespace and the C# API in a different namespace should make everything very clear

File: src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+public record MyRecord(string Name, int Age);
+
+public class GenTests
+{
+  [Fact]
+  public void ShouldUseGenConfig()
+  {
+    var onlyString = "singleton string";
+    var config = GenX.Defaults.WithGenerator(Gen.FromValue(onlyString));
```
[the remark this answers is not in the document]
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs:14]: I guess this is because both `Hedgehog` and `Hedgehog.Linq` define type called `Gen`.
This is super confusing because `Gen` in `Hedgehog` is the actual generator, while `Gen` in `Hedgehog.Linq` is a "static class" with helper C#-friendly methods in it.

C# also seems to resolves types relatively to the current namespace hierarchy. 
So when you do `namespace` after `using`, your `namespace` seems to be "closer" scope and it searches the type `Gen` in the namespace, traversing up in its hierarchy. And it finds `Gen` in `Hedgehog` before it explores "global" usings.

Having different things named the same (`Gen`) in two different namespaces when you practically need both to do something more or less realistic is super confusing.

[the remark this answers is not in the document]
Comment (dharmaturtle) [reply src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs:14]: Sounds about right to me: https://stackoverflow.com/a/151560

[the remark this answers is not in the document]
Comment (TysonMN) [reply src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs:14]: Yes, I understand this now.  Thanks to both of you.

So we give them different names, right?

How about we rename `Hedgehog.Linq.Gen` to `Hedgehog.Linq.GenL`?

**Comment by TysonMN** [comment]
Your suggest to only have one `GenX` seems good.

I think one reason we moved the `CompileName` attributes to their own file was to make things easier when adding support for the Fable compiler.  We are not doing that here (and probably never will because of all the reflection).

**Comment by AlexeyRaga** [comment]
OK, @TysonMN please have a look again.
I have moved helper functions into `Hedgehog.Linq` and made some extension methods in there.

For those functions that cannot be made extensions, noticeably `GenX.auto`, `GenX.autoWith` and `GenX.defaults` I have removed `CompileName` attribute because you were hesitant about it.
It'll look a bit unnatural from C# perspective, but I think that it is minor.

File: src/Hedgehog.Experimental/Gen.fs (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental/Gen.fs]: I put C# extension methods into `Hedgehog.Linq` namespace

File: src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs:15]: In `Hedgehog`? 
I _think_ that the effort there is to make `Hedgehog.Linq` to be self-sufficient. So C# devs would just reference that namespace and never resort to `Gen` in `Hedgehog`. 

That's fair.
But the fact that `Gen` exists in both namespaces, IDE sometimes gets confused and the mess starts.

TBH, because C# devs should probably never need to access `Hedgehog.Gen` I would use the same trick that F# itself uses for `Option`, `Result`, etc. and added `CompileName` attribute on `Hedgehog.Gen`.

This way in F# everything would be as it is, but from C# `Hedgehog.Gen` type would have a different name and there won't be any confusion.

But I think that having `GenL` in API isn't the nicest user experience...

[the remark this answers is not in the document]
Comment (TysonMN) [reply src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs:15]: > I _think_ that the effort there is to make `Hedgehog.Linq` to be self-sufficient. So C# devs would just reference that namespace and never resort to `Gen` in `Hedgehog`.

Oh, yes.  That is correct.  I had forgot that when I was working with you code earlier today.

> TBH, because C# devs should probably never need to access `Hedgehog.Gen` I would use the same trick that F# itself uses for `Option`, `Result`, etc. and added `CompileName` attribute on `Hedgehog.Gen`.
> 
> This way in F# everything would be as it is, but from C# `Hedgehog.Gen` type would have a different name and there won't be any confusion.

IIRC, we introduced this split after support for Fable was added.  Fable can't handle the `CompileName` attribute, so we had to wrap every `CompileName` attribute in compiler directives.  It made the code very verbose.  I found it very hard to "just" read the underlying F# code.

> But I think that having `GenL` in API isn't the nicest user experience...

I don't like this either.  Just threw out the idea to see what you thought.

File: src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
+namespace Hedgehog.Experimental.CSharp.Tests;
+
+using Hedgehog.Linq;
+using static Hedgehog.Linq.Property;
+using FluentAssertions;
```
Comment (TysonMN) [inline src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs:5]: Visual Studio 2022 (Community Edition) has nine messages about how to improve this code.  Three of them are to move these `using` declarations before the `namespace`.  I tried to do that before but ran into the problems (as mentioned before/elsewhere).

Can you address these nine messages?

Maybe it would help if this `namespace` were `Hedgehog.Linq.Tests`.

File: src/Hedgehog.Experimental.CSharp.Tests/Usings.cs [hunk 4]
```diff
@@ -0,0 +1 @@
+global using Xunit;
```
Comment (TysonMN) [inline src/Hedgehog.Experimental.CSharp.Tests/Usings.cs:1]: Why this instead of `using Xunit;` in `GenTests.cs`?

**Review by TysonMN — CHANGES_REQUESTED** [review]

**Comment by TysonMN** [comment]
> For those functions that cannot be made extensions, noticeably `GenX.auto`, `GenX.autoWith` and `GenX.defaults` [...]

Why can't `auto` and `autoWith` me made into extension methods?  (Since `defaults` is a value and not a function, I understand why it cannot be made into an extension method.)

File: src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs (hunk 3 above)
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental.CSharp.Tests/GenTests.cs:5]: Unlikely. I don't have Visual Studio and I don't run (or have access to) Windows.

File: src/Hedgehog.Experimental.CSharp.Tests/Usings.cs (hunk 4 above)
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental.CSharp.Tests/Usings.cs:1]: No reason, I just created a new test project and .NET (or whatever) created me this as a template. Perhaps this is the modern way in .NET/C#...

**Comment by AlexeyRaga** [comment]
> Why can't auto and autoWith me made into extension methods? 

Extension methods on what? These are static functions, and extension methods extend "values of type". There is no `this` for it to extend on.

File: src/Hedgehog.Experimental/Gen.fs [hunk 5]
```diff
@@ -238,6 +236,27 @@ module GenX =
...
+      arr |> Array.sortInPlaceBy f
+      (arr[0], arr[1], arr[2]))
+
+  /// Generates a 3-tuple with elements sorted by a given predicate.
+  let sortedBy4 (f : 'a -> 'k) (g : Gen<'a * 'a * 'a * 'a>) =
+    g |> Gen.map (fun (x1, x2, x3, x4) ->
+      let arr = [| x1; x2; x3; x4 |]
+      arr |> Array.sortInPlaceBy f
+      (arr[0], arr[1], arr[2], arr[3]))
+
```
[the remark this answers is not in the document]
Comment (AlexeyRaga) [reply src/Hedgehog.Experimental/Gen.fs:259]: Easier for me to remove, I'll drop them.

File: src/Hedgehog.Experimental/Gen.fs [hunk 6]
```diff
@@ -384,7 +384,7 @@ module GenX =
     |> AutoGenConfig.addGenerator uri
 
   module internal MultidimensionalArray =
-    
+
```
Comment (TysonMN) [inline src/Hedgehog.Experimental/Gen.fs:387]: Please revert this change since it is the only change in this file

**Review by TysonMN — CHANGES_REQUESTED** [review]

File: src/Hedgehog.Experimental/Gen.fs (hunk 6 above)
Comment (TysonMN) [reply src/Hedgehog.Experimental/Gen.fs]: Done by me
